[1단계 - 장바구니] 데이지(김유경) 미션 제출합니다.

kimyou1102 opened the pull request:
안녕하세요 마르코~~ 만나게되어서 반갑습니다. 이번 미션 잘부탁드립니다!!

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] RTL 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?

  - 배포 링크 기입: **https://kimyou1102.github.io/react-shopping-cart/**

- [ ] 리뷰어가 장바구니 추가를 쉽게 할 수 있도록 Curl 명령어를 DM 으로 전달해주세요.

  ```
  curl -X 'POST' \
  '<url>' \
  -H 'accept: */*' \
  -H 'Authorization: Basic <토큰>' \
  -H 'Content-Type: application/json' \
  -d '{
  "productId": 1,
  "quantity": 1
  }'
  ```

### 리뷰 요청 & 논의하고 싶은 내용
### 구조
![image](https://github.com/user-attachments/assets/6cbc05ab-5869-4711-a9aa-a24c72e92ee6)

### 1) 상태 설계 의도
이번 미션에서 불필요한 상태를 만들지 말고, SSOT를 지키는 것에 초점을 맞췄습니다.

원본으로부터 추론될 수 있는 값들은 모두 파생상태로 관리하였고, 서버로부터 내려받은 데이터는 가공 없이 그대로 상태로 두어 중복 저장을 피했습니다.


| 필드명                          | 키 (key)                      | 타입                     | 출처   | 관련 원본                    |
| ------------------------------- | ----------------------------- | ------------------------ | ------ | ---------------------------- |
| 장바구니 리스트                 | `cartItem`                    | `CartItemTypes[]`        | 원본   |                              |
| 장바구니에서 선택된 체크박스     | `selectedCartId`              | `string[]`               | 원본   |                              |
| 개별 장바구니 선택 아이콘        | `isChecked`                   | `boolean`                | 파생   | `selectedCartId`             |
| 상품 개수를 나타내는 안내 글     | `selectedCartId.length`       | `number`                 | 파생   | `selectedCartId`             |
| 주문 금액                       | `totalPrice`                  | `number`                 | 파생   | `cartItem, selectedCartId`   |
| 배송비                          | `deliveryFee`                 | `number`                 | 파생   | `totalPrice`                 |
| 총 결제 금액                    | `totalPrice + deliveryFee`    | `number`                 | 파생   | `totalPrice, deliveryFee`    |
| 주문확인 버튼의 활성화 여부     | `disabled`                    | `boolean`                | 파생   | `cartItem, selectedCartId`   |
| 확인 페이지의 장바구니 종류개수  | `selectedCartType`            | `number`                 | 파생   | `selectedCartId`             |
| 확인 페이지의 상품개수          | `selectedCartItem`            | `number`                 | 파생   | `cartItem, selectedCartId`   |
| 확인 페이지의 총결제 금액       | `totalPrice`                  | `number`                 | 파생   | `totalPrice`                 |


### 2) 이번 단계에서 가장 많이 고민했던 문제
첫 렌더 시 cartItem이 들어오면 전체 체크박스를 활성화하려고 했는데, 이대로 두면 수량 변경 등으로 cartItem이 바뀔 때마다 다시 전체 선택이 적용되는 부작용이 있었습니다.

1. Ref로 첫 렌더만 구분하기
useRef로 “첫 렌더 여부”를 저장해서, 한 번만 전체 선택을 수행하려 했습니다.

하지만 오직 이 목적만을 위해 ref를 추가하는 게 오히려 불필요한 상태라는 느낌이 들었고, 다른 곳에서는 쓸 일이 없겠다는 단점이 있었습니다.

cartItem이 불어와지지 않았을때 즉, 0이었을때만 로딩이 끝났다는 표시를 해주면 이후로는 로딩이 true이니 useEffect에서 한번만 실행되게 됩니다.

2. isLoading 상태 도입
대신 isLoading 플래그를 도입해, cartItem이 빈 배열(로딩 전)인 동안만 로딩 중으로 간주하고, 그 이후에는 한 번만 전체 선택 로직을 실행하도록 변경했습니다.

이렇게 하면 “처음 데이터가 들어왔을 때”라는 의미를 좀 더 명시적으로 관리할 수 있고, 다른 컴포넌트나 UI에서도 로딩 상태를 재사용할 수 있다는 장점이 있습니다.
```
 const getCartItemData = async () => {
    try {
      setIsLoading(true);
      const response = await getShoppingCart();
      if (cartItem.length === 0) setIsLoading(false);
      setCartItem(response);
    } catch (e) {
      setError("데이터를 가져오는데 실패했습니다");
    }
  };

  useEffect(() => {
    getCartItemData();
  }, []);

  useEffect(() => {
    if (!isLoading && cartItem) {
      setSelectedCartId(cartItem.map((item) => item.id.toString()));
    }
  }, [isLoading, cartItem]);
```
데이터 리패칭이나 실시간 업데이트가 필요할 때는, isFetching 같은 상태를 추가해 함께 사용해 UI를 처리할 수 있을 것 같습니다. 이러한 흐름이 괜찮은지, 제가 생각하는 isFetching더한 확장법이 맞는 방법인지 마르코의 의견 궁금합니다!

### 3) 이번 리뷰를 통해 논의하고 싶은 부분

**1. 체크 박스 상태**
체크박스 상태 관리를 최대한 파생 상태로만 처리해 보았습니다. 체크박스를 클릭하면 selectedCartId 배열에 해당 아이디가 없을 때만 추가하도록 하고, 각 체크박스 컴포넌트에는 isChecked={selectedCartId.includes(id)} 형태로 전달합니다. 

일반적으로는 체크박스마다 별도의 value 상태를 1:1로 관리하곤 했는데, 이렇게 하면 오직 id 배열 하나만으로 체크 여부를 결정할 수 있어 상태 관리가 더 간결해집니다. 이런 설계가 적절할지, 아니면 놓치고 있는 단점이 있을지 궁금합니다!

**2. SSOT**
낙관적업데이트를 하려면 서버에 요청을 보내기전에 화면을 바꾸어야해서 값을 저장하고 바꾸야 합니다. 그렇게 하면 서버의 상태와 이곳의 상태가 달라져서 SSOT 원칙에 어긋나지 않을지 고민입니다. 마르코의 의견이 궁금합니다!!

**3. MSW**
이번 미션에서는 msw를 모킹만을 위해 사용했습니다. msw는 서버가 완성되기 전에 사용하는 용도로도 쓸 수 있지만 테스트 환경의 복잡한 모킹을 하지 않고 사용될 수도 있다고 생각했습니다. 또한 테스트 환경에서 외부 API 호출 없이도 테스트가 일관되고 안정적으로 실행될 수 있어서 좋다고 생각합니다. 꼭 서버 대체용으로만 사용해야하는 걸까요? 마르코의 의견이 궁금합니다!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 클라이언트 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. MSW/Test

- 주요 기능을 적절히 정의하였나요?
- 주요 기능/예외에 대한 테스트가 충분히 이루어졌나요?

woowa-euijinkk committed:
chore: 의존성 제거

mun-kyeong committed:
docs: README.md에 구현 기능 목록 및 요구 사항 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: emotion 설치 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: Header 컴포넌트 및 스타일 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
chore: App.css 파일 삭제 및 스타일 리셋 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: Main 컴포넌트 및 스타일 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 장바구니 및 상품 컴포넌트 추가, 버튼 및 수량 선택기 구현

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: CheckBox 컴포넌트 및 스타일 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: Line 컴포넌트 및 스타일 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: CartProduct 및 CartProductContainer 스타일 및 구조 개선

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: CheckBox 컴포넌트의 스타일 import 수정 및 id prop 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: SelectAllLayout 추가 및 전체 선택 체크박스 구현

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: PaymentSummary 컴포넌트 추가 및 스타일 정의

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: SummaryRow 컴포넌트 추가 및 스타일 적용

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: PaymentSummary 컴포넌트의 레이아웃 및 스타일 개선

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 주문 확인 버튼 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 주문 확인 버튼 위치 변경

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 체크박스 컴포넌트에 onChange 핸들러 추가 및 CartProductContainer에서 선택 기능 구현

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 장바구니 API 기능 추가 (GET, POST, PATCH, DELETE)

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 장바구니 데이터 관리 훅 및 컴포넌트 통합

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 장바구니 관련 상태 관리 훅 제거 및 컴포넌트 통합

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: Toast 컴포넌트 및 스타일 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: Toast 컴포넌트 추가 및 오류 처리 기능 구현

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
fix: handleDelete 함수에서 삭제할 아이디를 동적으로 처리하도록 수정

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
refactor: maxQuantity 속성 제거 및 QuantitySelector에서 관련 코드 정리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
refactor :  장바구니 아이템 상태 끌어올리기

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: PaymentSummary  가격 계산 로직 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 버튼 스타일 및 상태에 따
른 활성화 로직 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
fix: 상태 관리 개선을 위한 useEffect 최적화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 장바구니가 비었을 때 표시할 EmptyShoppingCart 컴포넌트 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: react-router-dom 패키지 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 장바구니 페이지 및 스타일 컴포넌트 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 장바구니 페이지 및 스타일 컴포넌트 재구성

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

woowa-euijinkk committed:
docs: PR template

mun-kyeong committed:
feat: PageLayout 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 주문 확인 페이지 및 스타일 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: getTotalPrice 유틸리티 함수 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 라우팅 추가 및 장바구니 페이지 레이아웃 개선

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 장바구니 페이지에 CartProductList 추가 및 Footer 컴포넌트 구현

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: PageLayout 컴포넌트의 배경색 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
chore : vitest 설치 및 환경 세팅

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 장바구니 페이지에 대한 테스트 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: msw 패키지 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: MSW 설정 및 테스트 환경 구성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 제품 및 장바구니 데이터 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
fix: CartItemTypes의 quantity 타입을 number로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 장바구니 핸들러 및 API 엔드포인트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: 라우터에 basename 추가 및 장바구니 테스트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
chore: jest 설정 파일 삭제

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
fix: jest 타입 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
fix: ProductTypes에서 quantity 필드 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
feat: SummaryRow에 dataTestId 추가 및 장바구니 수량 변경 테스트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

mun-kyeong committed:
test: 장바구니 수량 변경에 대한 테스트 추가 및 getTotalPrice 테스트 파일 생성

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: CheckBox 컴포넌트에 dataTestId 추가 및 장바구니 전체 선택 기능에 대한 테스트 케이스 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: 주문확인 페이지에 총 결제금액 표시 형식 수정 및 테스트 추가

Co-authored-by: kimyou1102 <yoosk5485@naver.com>

mun-kyeong committed:
feat: package.json에 deploy 스크립트 추가 및 homepage 필드 업데이트

kimyou1102 committed:
Merge branch 'main' into step1

kimyou1102 committed:
docs: 구현 사항 체크표시

kimyou1102 committed:
chore: 배포 세팅

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+        <div css={titleBox}>
+          <p css={titleStyle}>장바구니</p>
+          {cartItem.length !== 0 && (
+            <p css={subTitleStyle}>현재 2종류의 상품이 담겨있습니다.</p>
```
몇 종류의 상품이 담겨져 있는지 보여줄 때, 값인 "2" 가 하드코딩 되어 있네요.
상태로 변경해주세요~!

kimyou1102 replied:
변경해야 했는데 놓쳤네요😱 반영하겠습니다!

kimyou1102 replied:
수정커밋: [fix: 하드코딩된 장바구니 종류 개수 상태로 변경](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/ba3e44f83743df3a3427f39447872e6e5875ad4c)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
컴포넌트 파일 명도 컴포넌트의 이름과 동일하게 PascalCase로 작성하는 것이 일반적인 컨벤션이에요~! 
파일만 봐도 어떤 컴포넌트를 정의하는지 명확하기 때문입니다.

kimyou1102 replied:
수정커밋: [refactor: 컴포넌트 파일명 컴포넌트명과 동일하게 PascalCase로 작성](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/2a4e6f88204c67d1d716ded06ef4972e97712bea)

wonsss commented on src/pages/orderConfirm/orderConfirm.tsx:
```diff
@@ -0,0 +1,38 @@
+} from "./orderConfirm.style";
+
+export function OrderConfirm() {
+  const { state } = useLocation();
```
useNavigate, useLocation 훅을 활용해 이전 페이지에서 해당 데이터를 전달받는 방식을 잘 사용해주셨네요!

다만, 현재는 직접 URL로 접근하면, 다음과 같이 에러가 발생합니다.
이전 페이지로부터 전달받을 것으로 예상된 state가 없는 경우(직접 URL로 접근하는 경우)에 대한 예외 처리도 해주시면 사용자 경험이 더 향상될 수 있습니다.
간단히는 이런 경우, '잘못된 접근입니다' 타이틀과 '장바구니로 돌아가기' 버튼 등을 배치할 수 있을 것 같아요.
![image](https://github.com/user-attachments/assets/36aec913-eb1a-42ae-9e21-866e877bd259)

kimyou1102 replied:
url으로 접근했을때 막는 것을 고려하지 않았네요. url 접근까지 디테일하게 대응할 수 있게 반영하겠습니다!

kimyou1102 replied:
수정커밋: [fix: confirm페이지 url로 접근시 오류나는 문제 해결](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/337604038fa40b637e9fbf7d68e1374a67157996)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+import { CartItemTypes } from "../../types/cartItem";
+import { getTotalPrice } from "../../utils/getTotalPrice";
+import { subTitleStyle, titleBox, titleStyle } from "./shoppingCart.style";
+import { Footer } from "../../components/layout/Footer/Footer";
```
import 경로를 간결하게 관리해보시는 것도 좋을 것 같습니다.
예를 들어, 절대 경로(alias)를 설정하거나, 각 폴더에 index.ts 파일을 추가해서 export를 정리하면, 동일한 폴더 내 모듈들을 하나의 경로에서 import할 수 있어 코드가 더 깔끔해집니다.
이는 유지보수성과 가독성을 높이는 데에도 도움이 됩니다.

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  return (
+    <PageLayout>
+      <Header>
```
PageLayout, Header, Main, Footer로 구성된 시맨틱한 네이밍의 컴포넌트 사용이 좋네요!

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+    <PageLayout>
+      <Header>
+        <p>SHOP</p>
+        {Boolean(error) && <Toast>{error && error}</Toast>}
```
`{error && error}` 은 앞선 조건과 중복 조건이므로, 간단히 `{error}` 로 표현 가능합니다.

kimyou1102 replied:
수정커밋: [refactor: 중복 조건 제거](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/2c7f6e076cc1aa4059a91a3d713444009c2fe703)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+  const handleConfirm = () => {
+    navigate("/confirm", {
+      state: {
+        selectedCartType: selectedCartId.length,
+        selectedCartItem: calculateCartItemQuantity(),
```
selectedCartType은 이름만으로는 의미가 다소 모호할 수 있습니다. 해당 값이 선택된 상품 종류의 개수를 의미한다면, selectedItemGroupCount처럼 Count 접미사를 활용해 역할이 드러나도록 네이밍을 개선하는 것이 좋습니다.

마찬가지로 selectedCartItem은 선택된 상품들의 총 수량을 의미하므로, selectedItemQuantity처럼 Quantity 혹은 Count가 붙는 이름이 더 명확합니다.

두 변수 모두 number 타입이기 때문에, 이치럼 정량적 의미가 드러나는 네이밍 컨벤션을 따르는 것이 유지보수나 협업 측면에서 더 이해하기 쉬운 코드를 만드는 데 도움이 됩니다.

kimyou1102 replied:
GroupCount가 붙으니 종류의 수량을 나타내서 훨씬 알기 쉬운 네이밍이 되는 거 같네요! 네이밍은 신경쓴다고 해도 항상 어려운거 같습니다...

kimyou1102 replied:
수정커밋: [refactor: 명확한 의미를 위해 네이밍 변경](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/76e6a02f2bcd214a519ebfd23767c79758c3625b)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+    });
+  };
+
+  const getCartItemData = async () => {
+    try {
+      setIsLoading(true);
+      const response = await getShoppingCart();
+      if (cartItem.length === 0) setIsLoading(false);
+      setCartItem(response);
+    } catch (e) {
+      setError("데이터를 가져오는데 실패했습니다");
+    }
+  };
+
+  const handleError = (errorMessage: string) => {
+    setError(errorMessage);
+  };
+
+  useEffect(() => {
+    getCartItemData();
+  }, []);
```
getCartItemData 함수는 데이터 패칭 + 로딩 상태 관리 + 에러 핸들링의 역할을 포함하고 있어, 현재 컴포넌트에서 UI 로직과 함께 섞여 있는 상태입니다. 이 경우, 관심사 분리 관점에서 별도의 커스텀 훅(useFetchCartItems 등)으로 분리하는 것이 유지보수성과 재사용성 측면에서 더 적절합니다.

예를 들어 useFetchCartItems 훅으로 분리하면, cartItems, isLoading, error 상태 관리, refetch 호출 함수를 모두 훅 내부에서 책임지게 할 수 있어, 컴포넌트는 더 깔끔한 형태로 UI에 집중할 수 있습니다.
```ts
// 예시
const { cartItems, isLoading, error, refetch } = useCartItems();
```

kimyou1102 replied:
데이터 페칭관련 로직을 훅으로 묶으니 확실히 재사용성과 유지보수성에 이점이 있을 거 같네요. 그리고 관련된 수정 사항이 있어도 훅만 수정하면 되니 훨씬 좋을 거 같습니다👍

kimyou1102 replied:
수정커밋: [refactor: 장바구니 페칭 관련 로직 useFetchCartItems훅을 사용하는 것으로 변경](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/04537e76c9fcffa53f682f6dfb3cee040b2f4028)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+  const [cartItem, setCartItem] = useState<CartItemTypes[]>([]);
+  const [error, setError] = useState("");
+  const [isLoading, setIsLoading] = useState(true);
+  const [selectedCartId, setSelectedCartId] = useState<string[]>([]);
```
selectedCartId는 단일 ID처럼 보이지만 실제로는 ID 배열 (string[])을 나타내기 때문에 복수형 표현이 더 적절해 보입니다.
또한, Cart는 도메인 상 전체 장바구니를 의미할 수 있어, 보다 구체적으로 ‘선택된 항목 ID들’을 나타내는 이름이 더 명확합니다.

ex) selectedItemIds

kimyou1102 replied:
수정커밋: [refactor: 배열을 담는 다는 의미를 잘 나타내기 위해 복수형 표현으로 네이밍 변경](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/d4c5e713bd73b8fbebcc74248a8a44a50a2a1a4f)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+      if (selectedCartId.includes(b.id.toString())) return a + b.quantity;
+      return a;
+    }, 0);
+  };
```
현재 getTotalPrice와 calculateCartItemQuantity는 모두 cartItems 배열에서 selectedCartId에 포함된 항목을 필터링한 뒤, 각각 금액과 수량을 계산하는 구조입니다.
두 함수 모두 입력값과 필터링 로직이 동일하므로, 하나의 함수에서 금액과 수량을 함께 반환하는 형태로 통합하는 것도 고려해볼 수 있습니다.
이렇게 하면 코드 중복을 줄이고, 반복 루프도 하나로 줄여 성능 면에서도 효율적입니다.

또한, 관련 있는 비즈니스 로직을 하나의 함수로 응집해둠으로써 재사용 및 유지보수에 유리합니다.
```ts
// 예시
const { totalPrice, totalQuantity } = getCartSummary({
  cartItems,
  selectedCartIds,
});
```

kimyou1102 replied:
유틸함수에서도 반복되는 부분인 것을 놓쳤네요 이렇게 합치니 관련된 로직끼리 응집성도 높아져서 더 좋은 거 같습니다!!

kimyou1102 replied:
수정커밋: [refactor: 중복되는 필터 로직 합치기 위해 getCartItemSummary 생성 및 적용](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/d6a70f462073a59d66a8233eaa39d78b3f59fd84)

MarcoJang123 replied:
잘 수정해주셨네요 👍

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+              ? "secondary"
+              : "primary"
+          }
+          disabled={selectedCartId.length === 0 || cartItem.length === 0}
```
selectedCartId.length === 0 || cartItem.length === 0라는 조건문을 명확하게 의미를 갖는 boolean 변수로 추출하면 가독성과 유지보수성이 좋아집니다. 
조건식이 UI 상태 결정 조건이라면 특히 더 의미 있는 변수명이 필요합니다.
 ```ts
// 예시
const shouldDisableButton = cartItem.length === 0 || selectedCartId.length === 0;
```

kimyou1102 replied:
수정커밋: [refactor: UI 변경을 위한 조건문을 변수로 빼기](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/07128d4d78c5fe61a2309c0045a4f11a138cf48e)

wonsss commented on src/components/CartProductContainer/CartProductContainer.tsx:
```diff
@@ -0,0 +1,99 @@
+    } else {
+      setSelectedCartId([...selectedCartId, id]);
+    }
+  };
```
현재 Checkbox의 onChange 핸들러에서 id === "select-all" 조건을 통해 전체 선택과 개별 선택을 구분하고 있습니다.
하지만 이 방식은 매직 스트링에 의존하고, 조건문 내부에 여러 로직이 섞여 있어 가독성과 유지보수성 측면에서 불리합니다.

대신, Checkbox 마다 핸들러를 분리하면 구조적으로 더 명확하고 선언적인 코드가 됩니다.
각 핸들러가 단일 책임을 가지므로 테스트나 리팩토링 시 유리합니다.
```ts
          <CheckBox
            isChecked={
              selectedCartId.length === cartItem.length && cartItem.length !== 0
            }
            dataTestId="select-all"
            id="select-all"
            onChange={handleAllCheck}
          />

                  <CheckBox
                    dataTestId={`select-${item.id}`}
                    isChecked={selectedCartId.includes(item.id.toString())}
                    onChange={handleCheck}
                    id={item.id.toString()}
                  />
```

kimyou1102 replied:
같은 체크박스여서 하나로 해야한다고 생각했는데 이렇게 하니 역할에 따라 더 분리가 되어서 추후 확장에도 좋겠네요!

kimyou1102 replied:
수정커밋: [refactor: 체크박스 선택 이벤트 핸들러를 전체선택과 개별체크 선택 함수를 분리](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/ec7e5e1efe98585ae7864f5783bfcc49bd6a0e1e)

wonsss commented on src/components/CartProductContainer/CartProductContainer.tsx:
```diff
@@ -0,0 +1,99 @@
+                  </Button>
+                </div>
+                <CartProduct
+                  key={item.id}
```
현재 map 함수에서 반환되는 최상위 요소인 `<div css={CartItemBox}>`에 key prop이 누락되어 있습니다. 대신 하위 요소인 CartProduct에 key가 잘못 지정되어 있네요.
React에서 리스트 렌더링 시 반복되는 요소의 최상위 DOM 요소에 고유한 key를 부여해야 성능 최적화 및 재렌더링 이슈를 방지할 수 있습니다.

![image](https://github.com/user-attachments/assets/b8c75a08-984e-4544-a223-10c77a8b44f5)

kimyou1102 replied:
수정커밋: [fix: map함수에서 반환되는 최상위 요소에 key값 추가](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/a0f6d4a38b938a4e9ad0fca4cf93e3adf06dc841)

wonsss commented on src/components/QuantitySelector/QuantitySelector.tsx:
```diff
@@ -0,0 +1,45 @@
+    await patchShoppingCart(cartId, quantity - 1);
+
+    onChange();
+  };
```
에러 처리 및 로딩 처리를 보강하면 좋을 것 같습니다.

API 호출 실패 시 사용자에게 피드백이 필요합니다.

사용자가 느린 인터넷 환경에서 수량 변경 버튼을 빠르게 여러 번 누를 경우,
이전 API 응답이 오기 전에 새로운 patchShoppingCart 요청이 중복으로 발생할 수 있습니다.

이러한 불필요한 요청을 방지하고 UX를 개선하기 위해,
patchShoppingCart API의 로딩 상태를 기준으로 버튼을 disabled 처리하는 방법을 고려해보시면 좋을 것 같습니다.

wonsss replied:
또는, 로딩 중 disabled 처리 방식 대신 낙관적 업데이트(Optimistic Update) 방식을 검토해볼 수도 있어요.
낙관적 업데이트는 사용자에게 먼저 UI 업데이트 후 API 요청함으로써 사용자에게 즉시 피드백 제공할 수 있어 UX가 더 좋을 수 있습니다. 다만, API 요청 실패 시 롤백 처리가 필요하여 구현 복잡도가 더 있습니다.

낙관적 업데이트 관련하여 이미 질문을 주셔서 이어서 답변드릴게요!

> 3. SSOT
낙관적업데이트를 하려면 서버에 요청을 보내기전에 화면을 바꾸어야해서 값을 저장하고 바꾸야 합니다. 그렇게 하면 서버의 상태와 이곳의 상태가 달라져서 SSOT 원칙에 어긋나지 않을지 고민입니다. 마르코의 의견이 궁금합니다!!

여기서 SSOT 원칙의 기준은 장바구니 상태의 단일 출처는 서버여야 한다는 전제일 것 같아요.
낙관적 업데이트는 서버 응답 전에 화면을 먼저 바꾸기 때문에 일시적으로 클라이언트 상태와 서버 상태가 불일치하게 되는 것은 맞습니다. 이 시점만 떼어 보면 SSOT가 위배된 것처럼 보일 수 있어요. 
하지만 낙관적 업데이트는 UX 중심 설계에서 적절한 방식이며, 롤백 가능한 구조라면 SSOT를 훼손하지 않습니다.
조금 더 부연 설명하자면, 낙관적 상태는 임시 상태로서 일종의 캐시이며, 궁극적인 상태는 여전히 서버에 있다는 사실은 유지됩니다.

kimyou1102 replied:
수정커밋: [refactor: api 통신중일때 여러번 수량 변경 안되도록 막는 기능 구현](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/479b1e5aea6386ee2f1637e8f6a78ca93eddfd16)

wonsss commented on src/components/CartProductContainer/CartProductContainer.tsx:
```diff
@@ -0,0 +1,99 @@
+  cartItem: CartItemTypes[];
+  onChange: () => void;
+  onError: (message: string) => void;
+  selectedCartId: string[];
+  setSelectedCartId: (id: string[]) => void;
```
id의 타입으로 string보다 number 가 더 정확할 것 같아요. 이렇게 하면 toString() 같은 형변환을 하지 않아도 됩니다.

kimyou1102 replied:
id값으로 "selected-all" 값도 들어가야 해서 string타입으로 정했습니다. 

만약 number만 사용할 경우 number | string 유니온 타입을 써야 하고, 그에 맞는 추가적인 타입 검사 코드를 작성해야 해서 현재처럼 string 하나로 관리하고 있습니다!

wonsss commented on src/components/CartProductContainer/CartProductContainer.tsx:
```diff
@@ -0,0 +1,99 @@
+              <div css={CartItemBox}>
+                <Line />
+                <div css={CartItemHeader}>
+                  <CheckBox
+                    dataTestId={`select-${item.id}`}
+                    isChecked={selectedCartId.includes(item.id.toString())}
+                    onChange={handleCheckBox}
+                    id={item.id.toString()}
+                  />
```
> 1. 체크 박스 상태
체크박스 상태 관리를 최대한 파생 상태로만 처리해 보았습니다. 체크박스를 클릭하면 selectedCartId 배열에 해당 아이디가 없을 때만 추가하도록 하고, 각 체크박스 컴포넌트에는 isChecked={selectedCartId.includes(id)} 형태로 전달합니다.
일반적으로는 체크박스마다 별도의 value 상태를 1:1로 관리하곤 했는데, 이렇게 하면 오직 id 배열 하나만으로 체크 여부를 결정할 수 있어 상태 관리가 더 간결해집니다. 이런 설계가 적절할지, 아니면 놓치고 있는 단점이 있을지 궁금합니다!

 파생 상태만으로 체크 상태를 관리하는 설계를 적절히 잘 해주셨네요!
참고로, `selectedCartId.includes(id)`의 시간 복잡도는 O(n) 이라서 배열 개수가 많이 큰 경우에는 O(1) 성능의 Set 자료 구조(has 메서드)를 사용하는 것이 유리할 수 있습니다. 
그러나 이러한 장바구니 정도 개수에서는 유의미하지는 않을 것 같아서 참고만 해주셔도 돼요.

kimyou1102 replied:
시간복잡도를 위한 Set 자료구조 사용도 좋을 거 같네요! 덕분에 깊게 고려하지 못한 부분 알아갑니다!

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+    } catch (e) {
+      setError("데이터를 가져오는데 실패했습니다");
+    }
+  };
```
> isLoading 상태 도입
대신 isLoading 플래그를 도입해, cartItem이 빈 배열(로딩 전)인 동안만 로딩 중으로 간주하고, 그 이후에는 한 번만 전체 선택 로직을 실행하도록 변경했습니다.
이렇게 하면 “처음 데이터가 들어왔을 때”라는 의미를 좀 더 명시적으로 관리할 수 있고, 다른 컴포넌트나 UI에서도 로딩 상태를 재사용할 수 있다는 장점이 있습니다.
데이터 리패칭이나 실시간 업데이트가 필요할 때는, isFetching 같은 상태를 추가해 함께 사용해 UI를 처리할 수 있을 것 같습니다. 이러한 흐름이 괜찮은지, 제가 생각하는 isFetching더한 확장법이 맞는 방법인지 마르코의 의견 궁금합니다!

try/catch/finally 구문에서는 isLoading 상태를 finally 블록에서 false로 설정하는 것이 적절합니다. 요청이 성공하든 실패하든 비동기 작업이 종료된 시점을 명확하게 처리할 수 있기 때문입니다.

또한, 말씀하신 것처럼 isLoading은 첫 데이터 요청에 대한 로딩 상태를 의미하고, 이후 재요청(예: 수량 변경 후 재조회)이나 실시간 갱신 등에서는 isFetching과 같은 별도 상태를 도입해 관리하면 의미를 명확히 분리할 수 있습니다.
이렇게 분리하면 사용자 경험(UX)을 고려한 로딩 처리(예: 스켈레톤 대신 버튼 로딩 스피너)도 유연하게 대응할 수 있어 확장성과 유지보수성이 높아집니다.

kimyou1102 replied:
수정커밋:
[refactor: isLoading상태 변경 첫렌더링시 한번만 되도록 로직 수정](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/8307fccc95b3f8affd3efaed0cd20c29545e53c0)
[feat: 로딩중 UI 구현](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/5b8b1d54bb290afd7afcba0fca7189aa3a6e6a68)

wonsss commented on test/setup.ts:
```diff
@@ -0,0 +1,7 @@
+import "@testing-library/jest-dom";
+import "@testing-library/jest-dom/vitest";
+import { server } from "../src/mocks/server";
+
+beforeAll(() => server.listen());
+afterAll(() => server.close());
+afterEach(() => server.resetHandlers());
```
> 이번 미션에서는 msw를 모킹만을 위해 사용했습니다. msw는 서버가 완성되기 전에 사용하는 용도로도 쓸 수 있지만 테스트 환경의 복잡한 모킹을 하지 않고 사용될 수도 있다고 생각했습니다. 또한 테스트 환경에서 외부 API 호출 없이도 테스트가 일관되고 안정적으로 실행될 수 있어서 좋다고 생각합니다. 꼭 서버 대체용으로만 사용해야하는 걸까요? 마르코의 의견이 궁금합니다!

맞습니다! 테스트 전용 목적으로 쓰는 것도 매우 바람직합니다.
참고로, 스토리북에서도 msw 플러그인을 제공하고 있어서, 스토리북 내에서도 msw로 mocking할 수 있습니다.

wonsss commented on src/components/QuantitySelector/QuantitySelector.tsx:
```diff
@@ -0,0 +1,45 @@
+      />
+      <p data-testid="quantity-value">{quantity}</p>
+      <IconButton
+        imgUrl="./plus.png"
```
로컬에서 실행해봤을 때, 아이콘 버튼의 이미지들이 엑박으로 뜨고 있는데, 확인해주세요. 
![image](https://github.com/user-attachments/assets/1abee391-f3a2-47d2-ae73-92a199263e1d)

그리고 배포된 페이지 주소가 없는데 이 부분도 PR 본문에 추가 부탁드려요.
![image](https://github.com/user-attachments/assets/3e9fb07c-8dd9-43f2-97af-1f2ed39be126)

kimyou1102 replied:
수정커밋: [fix: 경로 문제 수정](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/da2f7ea88b0a618b130888ff177ad2c4c7b6fcc5)

wonsss commented on public/mockServiceWorker.js:
MSW의 이름에 포함되어 있는 service worker란 중요한 웹표준 기술 중 하나입니다.
시간적 여유가 되시면, 학습 차원에서 service worker에 대해 한번 알아보시면 웹 기술에 대해 더 깊이 이해하는 데 도움이 되실 것 같습니다.

wonsss commented on test/getTotalPrice.test.ts:
```diff
@@ -0,0 +1,20 @@
+        })
+      ).toBe(totalPrice);
+    }
+  );
```
선택된 장바구니 아이템에 따라 계산되는 파생 상태는 단순히 ‘총 금액’ 외에도 ‘총 수량’, ‘배송비’, ‘총 상품 종류 수’ 등으로 확장될 수 있습니다.
이러한 계산 로직들을 하나의 유틸 함수로 응집시켜 관리하면 관련 로직의 재사용성과 유지보수성이 향상될 수 있으며, 테스트 코드도 함께 보강하면 다양한 조합에 대한 신뢰성 높은 검증이 가능해질 것 같습니다.

wonsss commented on test/getTotalPrice.test.ts:
```diff
@@ -0,0 +1,20 @@
+    [["7124", "7161", "7457"], 114800],
+    [["7124", "7161"], 14800],
+    [["7457"], 100000],
+  ])(
```
it.each를 사용하여 다양한 선택 케이스에 대해 명확하게 검증하고 있어 좋습니다.

이를 활용하여, 빈 배열일 때 등의 경계값 케이스 추가를 고려해보시면 좋을 것 같습니다.

kimyou1102 replied:
수정커밋: [test: 경계값 테스트 추가](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/a761fd8f30aada46292765802933e67e22b4e080)

wonsss commented on test/shoppinCart.test.tsx:
```diff
@@ -0,0 +1,174 @@
+      <MemoryRouter>
+        <ShoppingCart />
+      </MemoryRouter>
+    );
```
반복되고 있어 유틸 함수로 분리하면 좋을 것 같습니다.

kimyou1102 replied:
수정커밋: [test: ShoppingCart 테스트 중복 제거를 위해 renderComponent 헬퍼 함수 추가](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/12c16dceb829b4fd9c93352b0eccec6d87e63aa2)

wonsss commented on test/OrderConfirm.test.tsx:
```diff
@@ -0,0 +1,43 @@
+afterAll(() => server.close());
+afterEach(() => server.resetHandlers());
+
+describe("주문확인 페이지 테스트", () => {
```
사용자가 직접 URL /confirm으로 진입했을 때(location.state가 누락된 경우)에 대한 예외 케이스에 대해서도 테스트를 추가해볼 수 있을 것 같습니다.

kimyou1102 replied:
수정커밋: [test: 확인 페이지 url로 직접 진입시 에러 화면 테스트 작성](https://github.com/woowacourse/react-shopping-cart/pull/358/commits/ee6c53d0acf868cb54f26a3e406b2f00ff2e4057)

wonsss commented on src/pages/shoppingCart/shoppingCart.tsx:
```diff
@@ -0,0 +1,114 @@
+    return cartItem.reduce((a, b) => {
+      if (selectedCartId.includes(b.id.toString())) return a + b.quantity;
+      return a;
+    }, 0);
+  };
+
+  const handleConfirm = () => {
+    navigate("/confirm", {
+      state: {
+        selectedCartType: selectedCartId.length,
+        selectedCartItem: calculateCartItemQuantity(),
+        totalPrice,
+      },
+    });
+  };
+
+  const getCartItemData = async () => {
+    try {
+      setIsLoading(true);
+      const response = await getShoppingCart();
+      if (cartItem.length === 0) setIsLoading(false);
+      setCartItem(response);
+    } catch (e) {
+      setError("데이터를 가져오는데 실패했습니다");
+    }
+  };
+
+  const handleError = (errorMessage: string) => {
+    setError(errorMessage);
+  };
+
+  useEffect(() => {
+    getCartItemData();
+  }, []);
+
+  useEffect(() => {
+    if (!isLoading && cartItem) {
+      setSelectedCartId(cartItem.map((item) => item.id.toString()));
+    }
+  }, [isLoading, cartItem]);
```
코드 전체적으로 커스텀훅으로 분리한 코드가 아직 없는 것 같아요.
> 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)
위와 같은 요구사항처럼, 상태 관리 로직의 책임을 적절히 응집/분리하는 것은 가독성/유지보수성/테스트성 등의 측면에서 중요합니다.
여기서는 예를 들어, useCartData, useCartSelection, getCartSummary 등으로 커스텀 훅과 유틸을 분리하는 것이 유지보수성과 테스트성을 높이는 방향입니다.

kimyou1102 replied:
미션을 구현하는 것을 최우선으로 두다보니 시간 부족으로 커스텀훅을 활용하지 못한 거 같습니다. 리팩토링을 단계에서 커스텀훅을 활용하여 관심사 분리와 유지보수성을 신경쓰도록 하겠습니다!

MarcoJang123 replied:
네! 좋습니다. 다음 단계에서 더 해보면 좋을 것 같아요.

wonsss reviewed (changes requested):
안녕하세요! 데이지 😄 
1단계 장바구니 미션 수고 많으셨습니다.
좋은 내용의 질문들 잘 해주셨고, 코드에서도 고민을 많이 하신 점이 느껴져서 좋았습니다. 👍
리뷰 반영 후 재리뷰 요청주세요. 감사합니다!

kimyou1102 committed:
fix: 하드코딩된 장바구니 종류 개수 상태로 변경

kimyou1102 committed:
refactor: 컴포넌트 파일명 컴포넌트명과 동일하게 PascalCase로 작성

kimyou1102 committed:
refactor: 외부에서 스타일 주입 가능하도록 Button 컴포넌트의 props로 customCss 추가

kimyou1102 committed:
fix: confirm페이지 url로 접근시 오류나는 문제 해결

- location으로 받아오는 state 유무에 따라 조건부 렌더링 추가

kimyou1102 committed:
refactor: 중복 조건 제거

kimyou1102 committed:
refactor: 명확한 의미를 위해 네이밍 변경

kimyou1102 committed:
feat: useFetchCartItems 훅 생성

kimyou1102 committed:
refactor: 장바구니 페칭 관련 로직 useFetchCartItems훅을 사용하는 것으로 변경

kimyou1102 committed:
refactor: 배열을 담는 다는 의미를 잘 나타내기 위해 복수형 표현으로 네이밍 변경

kimyou1102 committed:
refactor: 중복되는 필터 로직 합치기 위해 getCartItemSummary 생성 및 적용

kimyou1102 committed:
refactor: UI 변경을 위한 조건문을 변수로 빼기

kimyou1102 committed:
fix: map함수에서 반환되는 최상위 요소에 key값 추가

kimyou1102 committed:
refactor: 체크박스 선택 이벤트 핸들러를 전체선택과 개별체크 선택 함수를 분리

kimyou1102 committed:
fix: 선택된 개수가 0개 이상일때 전체선택 클릭시 전체 미선택 되는 오류 해결

- 선택된 id들의 배열의 길이가 cartIitems의 길이와 같을때만 빈배열로 변경되도록 수정

kimyou1102 committed:
fix: 경로 문제 수정

kimyou1102 committed:
test: getCartItemSummary 테스트 작성

kimyou1102 committed:
test: 변경된 유틸함수 및 state명에 맞게 테스트 변경

-       selectedCartType -> selectedItemGroupCount
- getTotalPrice 함수 -> getCartItemSummary 함수

kimyou1102 committed:
refactor: 불필요한 함수 제거 및 해당 함수 테스트 제거

kimyou1102 committed:
test: 경계값 테스트 추가

kimyou1102 committed:
test: ShoppingCart 테스트 중복 제거를 위해 renderComponent 헬퍼 함수 추가

kimyou1102 committed:
test: 확인 페이지 url로 직접 진입시 에러 화면 테스트 작성

kimyou1102 committed:
feat: 확인 페이지에서 뒤로가기 기능 추가

kimyou1102 committed:
refactor: isLoading상태 변경 첫렌더링시 한번만 되도록 로직 수정

- isLoading이 getCartItemData가 동작할때마다 true로 변하긴 하는데 맨처음 값이 없을때만 false로 하는 다소 어색한 로직
- useEffect문 안에서 isLoading 상태를 변경

kimyou1102 committed:
refactor: api 통신중일때 여러번 수량 변경 안되도록 막는 기능 구현

kimyou1102 committed:
feat: Spinner 컴포넌트 생성

kimyou1102 committed:
feat: 로딩중 UI 구현

kimyou1102 commented:
안녕하세요 마르코! 마르코 덕분에 미션을 진행하며 생긴 고민들에 많은 도움이 되었고 꼼꼼힌 리뷰덕에 놓친 부분들을 많이 챙겨갈 수 있었습니다. 감사합니다!!😊

리뷰에 대한 답변들과 반영사항은 대댓글로 수정커밋을 남겨뒀습니다~!

MarcoJang123 reviewed (approved):
안녕하세요. 데이지 😄 
1단계 장바구니 미션 리뷰 꼼꼼히 잘 반영해주셨네요 👍 감사합니다.
1단계 목표에 충분히 잘 도달한 것 같아서, 여기서 마무리하고 2단계로 넘어가면 좋을 것 같아요.
2단계에서 뵈어요! 화이팅 💪

wonsss reviewed (approved):
안녕하세요. 데이지 😄 
1단계 장바구니 미션 리뷰 꼼꼼히 잘 반영해주셨네요 👍 감사합니다.
1단계 목표에 충분히 잘 도달한 것 같아서, 여기서 마무리하고 2단계로 넘어가면 좋을 것 같아요.
2단계에서 뵈어요! 화이팅 💪
